File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());

        vector<vector<int>> result;
        for(int i=0; i<nums.size()-2 && nums[i]<=0; i++) {
            if(i>0 && nums[i] == nums[i-1]) continue;
            int j=i+1;
            int k=nums.size()-1;
            int target=nums[i]*-1;
            while(j<k) {
                int sum = nums[j] + nums[k];
                
                if (sum < target) j++;
                else if(sum > target) k--;
                else {
                    result.push_back({nums[i], nums[j], nums[k]});
                    
                    j++; 
                    k--;

                    while(j<k && nums[j]==nums[j-1]) j++;
                    while(j<k && nums[k]==nums[k+1]) k--;
                }
            }
        }

        return result;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []

        for idx, i in enumerate(nums):
            if idx > 0 and i == nums[idx-1]:
                continue

            l = idx+1
            r = len(nums)-1

            sum = -i
            
            while (l < r):
                curr_sum = nums[l] + nums[r]
                if curr_sum > sum:
                    r-=1
                elif curr_sum < sum:
                    l+=1
                else:
                    result.append([i, nums[l], nums[r]])
                    r-=1
                    l+=1

                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
                    
                    while l < r and r < len(nums)-1 and nums[r] == nums[r + 1]:
                        r -= 1

        return result
        